README_zh.md

```diff
@@ -0,0 +1,252 @@
+```
+
+接受所有的对话，除了频道（注意，不一定只是一个规则。多个规则是从前向后应用）：
+
+```json
+"history": [
+    "all",
+    {"exclude": {"type": "channel"}}
+]
+```
+
+接受对话和群组中的所有媒体文件，但是群组中的媒体文件有 500MiB 的大小限制：
+
+```json
+"media": [
+    {"type": "user"},
+    {"type": "group", "media_max_size": "500M"}
+]
+```
+
+接受所有对话中的媒体文件和两个群组中的媒体文件，两个群组媒体文件的大小限制是 500MiB：
+
+```json
+"media": [
+    {"type": "user"},
+    {"only": [
+        {"title": "Group A"},
+        {"title": "Group B"},
+    ], "with": {"media_max_size": "500M"}}
+]
+```
+
+`only`规则也可以这么写：
+
+```json
+{"title": "Group A", "media_max_size": "500M"},
+{"title": "Group B", "media_max_size": "500M"}
+```
+
+#### 属性的规则
+
+```json
+{
+    "id": 123,
+    "title": "Name",
+    "username": "uname",
+    "type": "user",
+    "media_max_size": "500M"
+}
+```
+
+通过下列的属性来匹配聊天和文件：
+
+* `id` 可以从 [聊天列表](#聊天列表) 获取。
+* `title` 对于用户来说是用户名。
+* `type` 可以是 `"user"`, `"group"` 或 `"channel"`;
+* `media_max_size`仅会被用在 `config.media` 并且必须类似这种形式 `"500M"`, `"500K"` 或 `"500"` (字节).
+
+#### 排除规则
+
+```json
+{"exclude": "inner rule"}
+```
+
+将会从聊天/文件的匹配规则中排除掉，即使他们前面已经被匹配上了。
+
+#### 列表规则
+
+```json
+["rule0", "rule1", "more rules"]
+```
+
+一个一个地应用列表中的规则。（前面也提到过，从前往后地应用）
+
+#### Only 规则
+
+```json
+{"only": "only-rule", "with": "with-rule"}
+```
+
+仅在`only-rule`规则匹配到时才会使用`with-rule`规则。
+
+#### All 规则
+
+```json
+"all"
+```
+
+匹配一切。
+
+#### None 规则
+
+```json
+"none"
+```
+
+什么都不匹配。
+
+### 聊天列表
+
+`tg_history_dumper -list-chats`
+
+将聊天以 `<type> <id> <title> (<username>)` 格式输出。
+对于用户来说，`Title `就是用户名。
+
+如果聊天没有匹配`config.history`规则，那么输出命令行日志时改行将会变灰。
+
+### 命令行参数
+
+一些参数将会覆盖`config`中的值。
+
+比如`-chat='Some Chat'`将会覆盖`config.history`中的对应值，这意味着数据更新时将只会处理来自`Some Chat`的。
+
+```
+$ tg_history_dumper --help
+Usage of tg_history_dumper:
+  -app-hash string
+      app hash
+  -app-id int
+      app id
+  -chat string
+      被备份的聊天的标题，将会覆盖 config.history
+  -config string
+      配置文件的路径 (默认 "config.json")
+  -debug
+      展示 debug 相关的日志信息。
+  -debug-tg
+      展示 debug TGClient 的日志信息。
+  -list-chats
+      列出来所有可用的聊天
+  -out string
+      备份目录的路径，将会覆盖 config.out_dir_path
+  -session string
+      session 文件路径，将会覆盖 config.session_file_path
+  -socks5 string
+      socks5 配置，包括 地址:端口，将会覆盖 config.socks5_proxy_addr
+```
+
+## 格式
+
+## 消息
+
+All messages are saved as JSON Lines (aka jsonl) to file `history/<id>_<title>`. Dumper searches directories only by id and renames folder when title is changed.
+
+所有的消息将会以 JSON 数据存储在 `history/<id>_<title>`，此工具将会只会通过 id 来搜索文件夹，并且当 `title` 变化时重命名文件夹。
+
+每个 JSON 对象都会有一个 `"_" 属性，`最外层的对象还有一个 `"_TL_LAYER"` 字段，其值是 API 的版本号。比如：
+
+```json
+{"Date":1601491406,"Message":"Hello World!","PeerID":{"ChannelID":1261507434,"_":"TL_peerChannel"},"_":"TL_message","_TL_LAYER":119}
+```
+
+（为了方便阅读，把其他很多属性都删除掉了）
+
+## 同级数据
+
+相关的用户和聊天会被分别存储在 `history/users` 和 `history/chats`中。每个文件都是一个带有基础用户/聊天数据（如用户名，标题等）的 JSON。
+
+不仅在遇到新的同级数据时添加一行新数据，而且在现有同级数据（例如标题）从先前的转储中更改后也添加一行新数据。 因此，相同的用户/聊天可能会多次出现。 每个ID的最后一条记录是最新的。
+
+> 这段不太理解，所以翻译的有点生硬，建议看英文原文。
```

Comment:
Original:
> Lines are added not only when new peer is encountered but also when existing peer data (title for example) has changed from previous dump. So same users/chats may appear multiple times there. The last record for each id is the most recent one.


I mean, `history/users` and `history/chats` files store changes in user and chat fields over time.

When started, the dumper first scans these files and remembers last values for each `id`. It is "the data from previous dump".

Then, each time the dumper encounters a peer (i.e. user, channel or group), it checks the peer fields. If at least one field has changed, new line is added to `users` (or `chats`) file.

Dumper may "encounter" peer when dumping that chat directly, when peer's message was forwarded, when peer is mentioned/linked, etc.

For example, if dumper was run twice and between those runs some user has changed his `username` from `someUser` to `some_user`, the `history/users` file will look like this:

```
...
{"ID":123, "Username":"someUser", "FirstName":"Hi", ... "UpdatedAt":"old stamp"}
...
{"ID":123, "Username":"some_user", "FirstName":"Hi", ... "UpdatedAt":"new stamp"}
...
```

Reply:
Alright, got it.